lang/java/reef-utils/src/main/java/org/apache/reef/util/MultiAsyncToSync.java

```diff
@@ -0,0 +1,173 @@
+ */
+package org.apache.reef.util;
+
+import org.apache.reef.util.exception.InvalidIdentifierException;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Assists a class in presenting a synchronous interface that is implemented
+ * via asynchronous interfaces and events. When a method call is received
+ * by the interface, parameter values are captured and the initiation of
+ * asynchronous processing is encapsulated in a callable object. When
+ * {@code block()} is called with the internal interface identifier, the
+ * lock is taken, the asynchronous processing is initiated, and the caller
+ * is put to sleep. After all of the asynchronous processing is complete the
+ * caller is released with a call to {@code release()}.
+ */
+public final class MultiAsyncToSync {
+  private static final Logger LOG = Logger.getLogger(MultiAsyncToSync.class.getName());
+
+  private final ConcurrentLinkedQueue<ComplexCondition> freeQueue = new ConcurrentLinkedQueue<>();
+  private final ConcurrentHashMap<Long, ComplexCondition> sleeperMap = new ConcurrentHashMap<>();
+  private final long timeoutPeriod;
+  private final TimeUnit timeoutUnits;
+
+  /**
+   * Initialize a multiple asynchronous to synchronous object with a specified timeout value.
+   * @param timeoutPeriod The length of time in units given by the the timeoutUnits
+   *                      parameter before the condition automatically times out.
+   * @param timeoutUnits The unit of time for the timeoutPeriod parameter.
+   */
+  public MultiAsyncToSync(final long timeoutPeriod, final TimeUnit timeoutUnits) {
+    this.timeoutPeriod = timeoutPeriod;
+    this.timeoutUnits = timeoutUnits;
+  }
+
+  /**
+   * Put the caller to sleep on a specific release identifier.
+   * @param identifier The identifier required to awake the caller via the {@code release()} method.
+   * @param asyncProcessor A callable object which returns {@code Boolean} that initiates the asynchronous
+   *                       processing associated with the call. This will occur inside the condition lock
+   *                       to prevent the processing from generating the signal before the calling thread
+   *                       blocks. {@code asyncProcessor.call()} should return true on success.
+   * @return A boolean value that indicates whether or not a timeout or error occurred.
+   * @throws InterruptedException The thread was interrupted while waiting on a condition.
+   * @throws InvalidIdentifierException The identifier parameter is invalid.
+   * @throws Exception The callable object referenced by the asyncProcessor parameter threw an exception.
+   */
+  public boolean block(final long identifier, final Callable<Boolean> asyncProcessor) throws Exception {
+    boolean error;
```

Comment:
use standard naming for Boolean variables, e.g. `hasError`, `isInterrupted`, etc.